### DISCUSSION TITLE
Where are Linux arm64 parsers sourced from? How to build from source and install tree-sitter?

### DISCUSSION THREAD

**Discussion by unphased** [nvim-treesitter/nvim-treesitter#4414] [Q&A] [answered]
tl;dr Skip to Questions section below!

### Overview 

I've had a great experience with treesitter (as well as easy peasy setup with `nvim-treesitter`) but encountered something today that's been difficult to track down: a difference in parsing behavior on different platforms.

I'm working with a GitLab CI YAML file and noticed that it properly detects and parses `bash` lanuage syntax inside the text entries on my M1 Mac and Ubuntu 20.04 environments, but now I brought the same config over to a linux-arm64 (Ubuntu 22.04 aarch64) VM and Treesitter is not treating the string values as bash language here. I checked and confirmed that via TSPlayground. 

I was able to find where the TS parser shared objects live on my system. They're at `~/.local/share/nvim/lazy/nvim-treesitter/parser`. But in particular I am having trouble finding information about what source was used to actually generate these parsers. Since these are .so's and specific to the platform, I have no way to compare them against each other in this form.

I'll do some more checks on my Raspberry Pi for another linux arm environment (won't be arm64 though, but still arm) presently.

I can't quite tell without digging in further whether these parser .so's are generated locally (and if so what source they're generated from) or downloaded. Although the nearly instant response to auto-install these makes me think it's fetched from the internet, I think these are probably also really fast to compile locally as well. Not sure if this supports this idea of the parsers being fetched from the internet, but I also saw this from checkhealth on this linux-arm64: ```WARNING `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)``` Presumably TSInstall fetches them from the internet. So where then does it fetch from?

In order to figure out if tree-sitter proper is maybe the culprit, I'm currently still tracking down the tree-sitter executable in the problem environment to find out what version it is. I ran the build again (see below for exactly how I'm building) and the only instance of tree-sitter or treesitter in the output is the dir `/share/nvim/runtime/lua/vim/treesitter` which just contains a bunch of lua files. This makes me believe that this environment is actually just entirely missing tree-sitter, further supporting the notion that the parser in question was fetched from the internet and potentially just outdated. On my mac I was able to track it down to this location: `/opt/homebrew/Cellar/tree-sitter/0.20.7/bin/tree-sitter`. This is a 12MB executable and a dependency that got installed alongside neovim by homebrew.

### Questions

1. parsers are fetched from the internet it seems. Is there a way to configure nvim-treesitter to build them from parser source? I found a 2 year old discussion on this topic asking in the other direction: https://github.com/nvim-treesitter/nvim-treesitter/discussions/1594 and it seemed like back then this was the default. How can I opt back into that behavior, since it probably might resolve this issue?
2. The assumptions in the above question (1) lead me to expect that even once I fix my neovim build from source process (again see below) to include tree-sitter along with the neovim installation, it may make `tree-sitter` available for `:TSInstallFromGrammar`, but the way things work now it will not be used to prepare parsers for nvim-treesitter auto-install. How can I automate it so I get the latest yaml parser on linux arm64? I believe this question might be a duplicate of question 1.
3. I looked in the build logs from the below build (`DOCKER_BUILDKIT=1 docker build --progress=plain --no-cache -o out .`) and I can see treesitter clearly being built but it's not being installed. Anyone know the incantation to get it to be installed? Does it make sense to try to do this? It'd be nice to have a quick way to update parsers to quickly test modifications on them. As it stands this is the biggest flaw with my build/install method right now. My suspicion is that the intended way to do this is just install tree-sitter independently on the system, which I will try next. ~*But if this is the case, and per above points, why is tree-sitter even needed for the neovim source build at all?*~ Update: It makes more sense now that I know that `cargo install tree-sitter-cli` might be the way to get `tree-sitter` and it looks like on closer inspection the nvim source build was just using `tree-sitter` library to generate the couple of built-in parsers (c, lua, viml, vimdoc). I think I'm satisfied with that.

### Appendix

The way I'm building and installing neovim is as follows:

Dockerfile
```
from ubuntu:20.04 as builder

RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get -y install ninja-build gettext libtool-bin cmake g++ pkg-config unzip curl git
RUN mkdir nvim
WORKDIR /nvim
RUN curl -L https://github.com/neovim/neovim/archive/refs/tags/nightly.tar.gz --output nvim-nightly.tar.gz && tar xf nvim-nightly.tar.gz && rm -rf nightly.tar.gz
WORKDIR /nvim/neovim-nightly

RUN make CMAKE_BUILD_TYPE=Release CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX=/opt/neovim"
RUN make install
RUN ls -laR /opt/neovim

# The export stage is used in this way to use this dockerfile to obtain a single built executable:
# DOCKER_BUILDKIT=1 docker build -o out .
# ^ The above assumes this dockerfile is named Dockerfile in its own dir
FROM scratch AS export-stage
COPY --from=builder /opt/neovim/ .
```

Build & Install shell script
```
#!/bin/bash

# Build the latest nvim downloaded from github. Instead of making a docker image dump the output in out/
DOCKER_BUILDKIT=1 docker build -o out .

# If nothing showed up (failed build?), abort. Heuristic here is to see if total dirsize greater than 20MB
if [ "$(du -s out | cut -f1)" -lt 20000 ]; then
    echo "Build output dir questionable, aborting"
    exit 1
fi

# Install in to system at predefined location which matches the configuration made during build
sudo rm -rf /opt/neovim
sudo mkdir -p /opt/neovim
sudo mv out/* /opt/neovim

# install the symlink if necessary
if [ ! -f /usr/local/bin/nvim ]; then
    sudo ln -s /opt/neovim/bin/nvim /usr/local/bin/nvim
fi
```

**Comment by amaanq** [comment]
1. It's in the docs, namely here you can point to a local repo: https://github.com/nvim-treesitter/nvim-treesitter#adding-parsers

If done via the tree-sitter cli, parsers are generated and placed at ~/.cache/tree-sitter/lib

I believe nvim-treesitter generates the shared object at `parsers/`, and will use the parser.c if generate_from_grammar is false, otherwise it'll run tree-sitter generate **and then** `cc parser.c -o parser.so` (while normally it'll just be this step, no generate), and move that parser to the `parsers/` dir.

I'm not sure exactly what your problem is without some more feedback/debug info

**Reply by unphased** [reply]
Thanks. I did find a yaml.so in `~/.cache/tree-sitter/lib` even though `tree-sitter generate --log` crashes. Maybe it's something specific to the logging that's crashing... 

is `~/.cache/tree-sitter/lib` documented somewhere? It really should be!

**Reply by unphased** [reply]
I found https://github.com/tree-sitter/tree-sitter/discussions/1711 but `-b` is not a valid flag to `generate`.

**Reply by unphased** [reply]
```
❯ ll ~/.local/share/nvim/lazy/nvim-treesitter/parser/yaml.so ~/.cache/tree-sitter/lib/yaml.so
-rwxrwxr-x 1 slu slu 192K Mar  2 02:53 /home/slu/.local/share/nvim/lazy/nvim-treesitter/parser/yaml.so
-rwxrwxr-x 1 slu slu 369K Mar  2 04:46 /home/slu/.cache/tree-sitter/lib/yaml.so
```
Looks promising!

**Reply by amaanq** [reply]
> is ~/.cache/tree-sitter/lib documented somewhere? It really should be!

Not that I'm aware of, a brief mention would be nice though. It's not that important imo because you can really just compile parser.c yourself, that's the beauty of tree-sitter.

> but -b is not a valid flag to generate.

What version?
What does `tree-sitter -V` output? 

I can confirm `tree-sitter g -b --libdir .` outputs the lang.so in the cwd for me, but I use a maintainer's fork with extra features. If you're interested... https://github.com/amaanq/tree-sitter-kdl/issues/1#issue-1559626795

**Reply by unphased** [reply]
```
❯ tree-sitter -V                      
tree-sitter 0.20.7
```

> you can really just compile parser.c yourself, that's the beauty of tree-sitter.

Can you elaborate on this? `parser.c` here is present and tracked in git in `tree-sitter-yaml`, and i was also confused earlier since I saw a git diff then as well even though I did not make the changes. Now I see this is proof that it's actually a generated file. Therefore it makes a bit of sense I can compile this into a `.so` for consumption. It's a big stack of assumptions (will admit, did not read every word of docs...) though.

Thanks for the assistance. Now:

- I put the new yaml.so in place (same spot as before) and I see no change in highlighting behavior. 
- So I need to track down **which version of tree-sitter-yaml** the yaml parsers on my mac and linux x64 system were built from to try to get that one! Totally lost how to find out that information. Also, no clue if `mv /home/slu/.cache/tree-sitter/lib/yaml.so /home/slu/.local/share/nvim/lazy/nvim-treesitter/parser/yaml.so` was a sensible thing to do.

**Reply by amaanq** [reply]
tree-sitter isn't updated to 0.20.9 on crates.io..unfortunate

Install either that 0.21.0 alpha fork or take the git version like so
`cargo install tree-sitter-cli --git https://github.com/tree-sitter/tree-sitter`

It'll still say v0.20.7 but that's because the manifest wasn't updated, I can confirm this works.

You won't notice a change in highlighting behavior from a regenerated parser, that'll come from queries FYI.

**Reply by amaanq** [reply]
To understand this better, the parser is generated from grammar rules. The grammar rules only generate a CST, the queries at queries/yaml/*.scm will determine how highlighting works. Notice how it captures groups or patterns of RULES to then link to a CAPTURE (`@type` for example). These captures are what enable your highlighting, so play around with that. Be careful though, if you're testing an updated custom query you will have a ton of error msgs if it's not done right, best way to do so is put it in the tree-sitter-yaml/queries dir and run tree-sitter test (tree-sitter t) to see if that complains first.

More info here https://tree-sitter.github.io/tree-sitter/syntax-highlighting

**Reply by unphased** [reply]
Thanks, this is helpful. Ah yeah, well clearly another one of the many assumptions i made along the way was that the parser that got installed was somehow out of date and not "smart enough" to interpret yaml strings as bash. A bridge too far...

**Reply by amaanq** [reply]
See injections.scm, maybe adding something like 
`(string) @bash` would work, but do check if the rule is named string in tree-sitter-yaml

**Reply by unphased** [reply]
Yeah I'm mildly familliar with how queries work since I was poking around with some higher level plugins like vim-matchup and I'm pretty happy with what can be accomplished in that regard (crafting queries) inside the tree-sitter Playground plugin. But yeah would be getting deep into this stuff if I wanted to define this stuff for new languages or other non-programming language type stuff. I don't like to design languages but a flexible and speedy parser in general I do actually have uses for!

**Comment by unphased** [comment]
Alright heres what I did first:
- `cargo install tree-sitter-cli`
- cloned `git@github.com:ikatyang/tree-sitter-yaml.git`
- checked out tag `v0.5.0`
- `RUST_BACKTRACE=full tree-sitter generate --log`
```
<snip>
[tree_sitter_cli::generate::build_tables::minimize_parse_table] split states 1540 1578 - successors for _r_acr_ctn are split: 1617 1628
[tree_sitter_cli::generate::build_tables::minimize_parse_table] split states 1540 1589 - successors for _r_acr_ctn are split: 1617 1629
[tree_sitter_cli::generate::build_tables::minimize_parse_table] split states 1578 1589 - successors for _r_acr_ctn are split: 1628 1629
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/slu/.cargo/registry/src/github.com-1ecc6299db9ec823/tree-sitter-cli-0.20.7/src/generate/build_tables/build_lex_table.rs:130:31
stack backtrace:
   0:     0xaaaab9e6a7d8 - std::backtrace_rs::backtrace::libunwind::trace::h4094424764928517
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0xaaaab9e6a7d8 - std::backtrace_rs::backtrace::trace_unsynchronized::h66a4d558a047d429
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0xaaaab9e6a7d8 - std::sys_common::backtrace::_print_fmt::he501a9dfcbc65602
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys_common/backtrace.rs:65:5
   3:     0xaaaab9e6a7d8 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h500c865c023970a8
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys_common/backtrace.rs:44:22
   4:     0xaaaab9e8c990 - core::fmt::write::h746b6837467f52c4
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/fmt/mod.rs:1208:17
   5:     0xaaaab9e646a8 - std::io::Write::write_fmt::h9c3773c135795f84
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/io/mod.rs:1682:15
   6:     0xaaaab9e6a5e4 - std::sys_common::backtrace::_print::haf80eda742d20a82
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys_common/backtrace.rs:47:5
   7:     0xaaaab9e6a5e4 - std::sys_common::backtrace::print::h1af712ffba8f7735
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys_common/backtrace.rs:34:9
   8:     0xaaaab9e6be6c - std::panicking::default_hook::{{closure}}::hd82c56e3724f8bd8
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:267:22
   9:     0xaaaab9e6bbb0 - std::panicking::default_hook::h1f2a94f3a1c1c1aa
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:286:9
  10:     0xaaaab9e6c4ec - std::panicking::rust_panic_with_hook::h01eb01a06bf17753
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:688:13
  11:     0xaaaab9e6c2cc - std::panicking::begin_panic_handler::{{closure}}::h657c9d34bb6d0855
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:579:13
  12:     0xaaaab9e6ac68 - std::sys_common::backtrace::__rust_end_short_backtrace::h0b2d3ba55b6e5445
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/sys_common/backtrace.rs:137:18
  13:     0xaaaab9e6c018 - rust_begin_unwind
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:575:5
  14:     0xaaaab9c087c0 - core::panicking::panic_fmt::h29641d0c529764f0
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:64:14
  15:     0xaaaab9c088a4 - core::panicking::panic_bounds_check::h4d2b52dba83c53ad
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/panicking.rs:147:5
  16:     0xaaaab9c9fb38 - <core::iter::adapters::map::Map<I,F> as core::iter::traits::iterator::Iterator>::next::hf7eb0aa30f160403
  17:     0xaaaab9caa068 - <alloc::vec::Vec<T> as alloc::vec::spec_from_iter::SpecFromIter<T,I>>::from_iter::h9bc6500389f4ae6c
  18:     0xaaaab9c591ac - tree_sitter_cli::generate::build_tables::build_lex_table::LexTableBuilder::add_state_for_tokens::he58b88a1e71158f2
  19:     0xaaaab9c57e00 - tree_sitter_cli::generate::build_tables::build_lex_table::build_lex_table::hcf41ad51d5c660d1
  20:     0xaaaab9cf77e0 - tree_sitter_cli::generate::build_tables::build_tables::h6709fd270b5f772a
  21:     0xaaaab9c68064 - tree_sitter_cli::generate::generate_parser_for_grammar_with_opts::h8982c154a22c9b62
  22:     0xaaaab9c676b4 - tree_sitter_cli::generate::generate_parser_in_directory::h8da3b2413f70515e
  23:     0xaaaab9c0eb98 - tree_sitter::run::he6fa0fcc666dc9dd
  24:     0xaaaab9c0b0d4 - tree_sitter::main::hec3b8b1f47b6eb7c
  25:     0xaaaab9c161b0 - std::sys_common::backtrace::__rust_begin_short_backtrace::heff64ee9000e946e
  26:     0xaaaab9c1f378 - std::rt::lang_start::{{closure}}::hd7cf98830e5598dc
  27:     0xaaaab9e5e4a8 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h529a25abf285dbab
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/core/src/ops/function.rs:606:13
  28:     0xaaaab9e5e4a8 - std::panicking::try::do_call::h660335bd09680bb9
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:483:40
  29:     0xaaaab9e5e4a8 - std::panicking::try::h94641ffd91114cda
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:447:19
  30:     0xaaaab9e5e4a8 - std::panic::catch_unwind::hb02f598a6506eb1a
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panic.rs:137:14
  31:     0xaaaab9e5e4a8 - std::rt::lang_start_internal::{{closure}}::h2bafc04d50d16864
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/rt.rs:148:48
  32:     0xaaaab9e5e4a8 - std::panicking::try::do_call::he27a0671832dd909
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:483:40
  33:     0xaaaab9e5e4a8 - std::panicking::try::h517b21015b9f6abd
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panicking.rs:447:19
  34:     0xaaaab9e5e4a8 - std::panic::catch_unwind::hc734f97b8d8dbcc6
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/panic.rs:137:14
  35:     0xaaaab9e5e4a8 - std::rt::lang_start_internal::h3fd169391cc4081c
                               at /rustc/d5a82bbd26e1ad8b7401f6a718a9c57c96905483/library/std/src/rt.rs:148:20
  36:     0xaaaab9c11a20 - main
  37:     0xffffb9dc73fc - <unknown>
  38:     0xffffb9dc74cc - __libc_start_main
  39:     0xaaaab9c08c30 - _start
  40:                0x0 - <unknown>
```
Something's not right.

**Reply by amaanq** [reply]
> RUST_BACKTRACE=full tree-sitter generate --log

curiously, without --log it compiles fine for me, but then I also get the panic with --log

Seems like a tree-sitter bug then

**Reply by unphased** [reply]
i guess return code when not running with `--log` doesn't indicate a failure and plus it does generate the artifacts i was after... so this is a red herring? Thank you for confirming @amaanq

**Reply by amaanq** [reply]
Yes. Feel free to submit a PR to tree-sitter.. :)

**Reply by theHamsta** [reply]
Maybe you could have a look onto grammar.json/node-types.json. first, node generates those files (like acompiler frontend) and then rust generates the C files from the JSON files

**Reply by unphased** [reply]
Awesome. Thanks for the informative summary!

**Comment by unphased** [comment, marked as answer]
I think I solved the mystery. I had never even opened up a bash file ever in nvim in this environment yet, so all software was apparently doing the reasonable and right thing. It makes perfect sense to me that nvim-treesitter would not have the "foresight" to auto-install bash upon opening a yaml file. Actually it's still a mystery how treesitter could even guess that the strings might want to be bash inside since YAML can be used for many different things. But however that stuff works, it really does work.

All I did to make it behave like I expect was open up a shell script which caused nvim-treesitter to auto-install the bash treesitter parser, then my yaml files and dockerfiles started showing bash inside like I wanted.

**Reply by unphased** [reply]
appreciate your quick input @amaanq and thank you for contributing to the nvim ecosystem!

**Reply by amaanq** [reply]
No problem :)
